core/sdk-core/src/test/java/software/amazon/awssdk/core/progress/DefaultProgressSnapshotTest.java

```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    public void remainingBytes_withoutTotalBytes_isEmpty() {
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(1L)
+                                                                  .build();
+        assertThat(snapshot.remainingBytes()).isNotPresent();
+    }
+
+    @Test
+    public void remainingBytes() {
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(1L)
+                                                                  .totalBytes(5L)
+                                                                  .build();
+        Assertions.assertEquals(4.0, snapshot.remainingBytes().getAsLong(), 0.0);
+    }
+
+    @Test
+    public void elapsedTime_withoutStartTime_isEmpty() {
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(1L)
+                                                                  .build();
+        assertThat(snapshot.elapsedTime()).isNotPresent();
+    }
+
+    @Test
+    public void elapsedTime() {
+
+        Instant startTime = Instant.now().minusMillis(100);
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(1L)
+                                                                  .startTime(startTime)
+                                                                  .build();
+        Duration expectedDuration = Duration.between(startTime, Instant.now());
+
+        Assertions.assertEquals(snapshot.elapsedTime().get().toMillis(), expectedDuration.toMillis(), 0.1);
+    }
+
+    @Test
+    public void averageBytesPer_withoutStartTime_isEmpty() {
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(1L)
+                                                                  .totalBytes(5L)
+                                                                  .build();
+        assertThat(snapshot.averageBytesPer(TimeUnit.MILLISECONDS)).isNotPresent();
+    }
+
+    @Test
+    public void averageBytesPer() {
+        DefaultProgressSnapshot snapshot = DefaultProgressSnapshot.builder()
+                                                                  .transferredBytes(100L)
+                                                                  .startTime(Instant.now().minusMillis(100))
+                                                                  .build();
+        Assertions.assertEquals(1.0, snapshot.averageBytesPer(TimeUnit.MILLISECONDS).getAsDouble(), 0.2);
+    }
+
+    @Test
+    public void estimatedTimeRemaining_withoutStartTime_isEmpty() {
```

Comment:
Can we use ParameterizedTest instead of write separate test methods?

Reply:
Thats a good idea. Much neater. Will do this